### PR TITLE
chore: add error boundaries across frontend when users don't have access

### PR DESCRIPTION
**Pull request by seaerchin**
## Problem
Previously, when we had an error, the users get shunted into the default error boundary which isn't very useful. this PR updates the layout of each individual page on our frontend so that we show the relevant error boundary.

## Solution
1. add a `PermissionBoundary` on each layout so that we perform permissions checking **prior** to rendering the page. this means, for example, that if page 234 doens't exist but the user has no access to it, they'll see the error screen for no permissions rather than the not found page. 
2. minor refactor so that we shift off the old pattern of using the `Can` from casl but instead, create a contextual `Can` that can read from the components appropriately

## Screenshots

https://github.com/user-attachments/assets/75b167f4-17ba-4d59-b32a-8e022ac4c45d

layouts after change 

https://github.com/user-attachments/assets/f83af47e-f756-4a97-baff-9f89a2dee075

error boundaries for all of our pages

### CODE REVIEW

File: apps/studio/src/features/dashboard/PermissionsErrorPage.tsx [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  title,
+  description,
+  buttonText,
+}: PermissionsErrorBoundaryProps) => {
+  const router = useRouter()
+
+  return (
+    <Center minH="$100vh">
+      <Stack alignItems="center" dir="vertical" gap="3rem">
+        <HeadScratch></HeadScratch>
```
Comment (adriangohjw) [inline apps/studio/src/features/dashboard/PermissionsErrorPage.tsx:22]: nit - `<HeadScratch />`

**Review by adriangohjw — APPROVED** [review]
> for example, that if page 234 doens't exist but the user has no access to it, they'll see the error screen for no permissions rather than the not found page. 

curious, any particular product/design reason why this is the case?

**Comment by seaerchin** [comment]
> > for example, that if page 234 doens't exist but the user has no access to it, they'll see the error screen for no permissions rather than the not found page.
> 
> curious, any particular product/design reason why this is the case?

thinking more on leaking of info - don't tell the user that it doesn't exist!

**Comment by adriangohjw** [comment]
Okay 👍 I think Google drive shows 404 but I don't think it's an issue
either way. Let's merge

**Comment by seaerchin** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/756?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#756** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/756?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#731** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/731?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#724** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/724?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#714** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/714?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#712** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/712?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#693** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/693?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#683** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/683?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#681** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/681?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @seaerchin and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by seaerchin** [comment]
### Merge activity

* **Oct 23, 3:34 PM EDT**: A user merged this pull request with [Graphite](https://app.graphite.dev/github/pr/opengovsg/isomer/756).

**Pull request merged**